sql/wsrep_sst.cc

```diff
@@ -1465,121 +1438,313 @@ std::string wsrep_sst_prepare()
 
-    if (NULL != proc.from())
-    {
-      proc.wait();
-    }
+  Parser_state ps;
+  if (ps.init(thd, thd->query(), thd->query_length()))
+  {
+    WSREP_ERROR("SST query: %s failed", query);
+    ret= -1;
+  }
+  else
+  {
+    PSI_statement_locker *parent_locker= thd->m_statement_psi;
+    thd->set_query_id(next_query_id());
 
-    if ((ret= proc.error()))
+    mysql_parse(thd, thd->query(), thd->query_length(), &ps);
+    if (thd->is_error())
     {
-      WSREP_ERROR("Try %d/%d: '%s' failed: %d (%s)",
-                  tries, max_tries, proc.cmd(), ret, strerror(ret));
-      sleep (1);
-    }
-    else
-    {
-      WSREP_DEBUG("SST script successfully completed.");
-      break;
+      int const err= thd->get_stmt_da()->sql_errno();
+      WSREP_WARN ("Error executing '%s': %d (%s)",
+                  query, err, thd->get_stmt_da()->message());
+      thd->clear_error();
+      ret= err;
     }
+    thd->m_statement_psi= parent_locker;
+    thd->end_statement();
+    close_thread_tables(thd);
+    delete_explain_query(thd->lex);
   }
 
-  return -ret;
+  thd->reset_query();
+  thd->cleanup_after_query();
+  return ret;
 }
 
-static void sst_reject_queries(my_bool close_conn)
+
+#define SST_USER_PREFIX "wsrep.sst."
+
+static std::string
+generate_user_name(const char* const node_name)
 {
-  WSREP_INFO("Rejecting client queries for the duration of SST.");
-  if (TRUE == close_conn) wsrep_close_client_connections(FALSE);
+  static size_t const max_user_name_length(32); // MySQL 8.0
+
+  struct timeval tv;
+  gettimeofday(&tv, NULL);
+  time_t time_now;
+  time_now = tv.tv_sec;
+  struct tm tm_now;
+  (void)gmtime_r(&time_now, &tm_now);
+  char timestamp[32];
+  strftime(timestamp, sizeof(timestamp), "%y%m%d_%H%M%S", &tm_now);
+
+  return (std::string(SST_USER_PREFIX) + timestamp + "_" + node_name).substr
+          (0, max_user_name_length);
 }
 
-static int sst_donate_mysqldump (const char*         addr,
-                                 const wsrep::gtid&  gtid,
-                                 bool                bypass,
-                                 char**              env) // carries auth info
+void
+wsrep_sst_cleanup_user(THD* const thd)
 {
-  char host[256];
-  wsp::Address address(addr);
-  if (!address.is_valid())
-  {
-    WSREP_ERROR("Could not parse SST address : %s", addr);
-    return 0;
+  int err(run_sql_command(
+          thd, "DELETE FROM mysql.user WHERE user LIKE '" SST_USER_PREFIX "%';"));
+  if (err && err != ER_NO_SUCH_TABLE) {
+    WSREP_WARN("Failed to clean up SST user(s): %d", err);
   }
-  memcpy(host, address.get_address(), address.get_address_len());
-  int port= address.get_port();
-  bool extra_args;
-  size_t const cmd_len= estimate_cmd_len(&extra_args);
-  wsp::string cmd_str(cmd_len);
+}
 
-  if (!cmd_str())
+static std::string
+generate_password(int size) {
+  // Password characters are limited, because we are using these passwords
+  // within a bash script.
```

Comment:
I don't think that diff is applicable here. The point here is that we want the password string to not contain any symbols that need to be escaped in bash, precisely so that we don't need to bother with escaping when exporting the variable.
As you mentioned yourself:
> To correctly SQL escape passwords, escaping \ first is
required. Then we need to escape ' in the password to
prevent it being treated as a end of SQL statement quote.

Why bother with characters that need escaping?